index_release.go

```diff
@@ -19,6 +20,95 @@ type IndexRelease struct {
 	Version     string      `yaml:"version"`
 }
 
+// CompileReleases takes indexReleases and collected version bundles and
+// compiles canonicalized Releases from them.
+func CompileReleases(logger micrologger.Logger, indexReleases []IndexRelease, bundles []Bundle) ([]Release, error) {
+	releases := buildReleases(logger, indexReleases, bundles)
+
+	deduplicateReleaseChangelog(releases)
+
+	return releases, nil
+}
+
+func buildReleases(logger micrologger.Logger, indexReleases []IndexRelease, bundles []Bundle) []Release {
+	bundleCache := make(map[string]Bundle)
+
+	// Create cache of bundles for quick lookup
+	for _, b := range bundles {
+		bundleCache[b.ID()] = b
+	}
+
+	var releases []Release
+
+INDEX_RELEASES:
+	for _, ir := range indexReleases {
+		release := Release{
+			active:    ir.Active,
+			timestamp: ir.Date.Format(indexReleaseTimestampFormat),
+			version:   ir.Version,
+		}
+
+		for _, a := range ir.Authorities {
+			b, found := bundleCache[a.BundleID()]
+			if !found {
+				logger.Log("level", "warning", "message", "IndexRelease v%s contains Authority with bundle ID %s that cannot be found from collected version bundles. Skipping.")
+				continue INDEX_RELEASES
+			}
+			release.bundles = append(release.bundles, b)
+		}
+
+		releases = append(releases, release)
+	}
+
+	return releases
+}
+
+// deduplicateReleaseChangelog removes duplicate changelog entries in
+// consequtive release entries.
+func deduplicateReleaseChangelog(releases []Release) []Release {
+	if len(releases) < 2 {
+		return releases
+	}
+
+	type LogState int
+	const (
+		New LogState = iota
+		Removed
+	)
+
+	sort.Sort(SortReleasesByVersion(releases))
+
+	prevChangelogs := make(map[string]LogState)
```

Comment:
This map with `LogState` might be unnecessary. I don't fully use it yet, but this was my first intuition.